### ISSUE TITLE
Compile error on Windows using MinGW

### ISSUE THREAD

**Issue by gamedevsam** [godotengine/godot#35022]
**Godot version:**
master branch

**OS/device including version:**
Windows 10 Pro v1909
Python v3.8.1
SCons  v3.1.2

**Issue description:**
```
TypeError: argument of type 'ListAction' is not iterable:
  File "C:\Users\Samuel\Dev\Godot\godot\godot\SConstruct", line 521:
    SConscript("modules/SCsub")
  File "C:\Users\Samuel\scoop\apps\scons\3.1.2\scons-local-3.1.2\SCons\Script\SConscript.py", line 660:
    return method(*args, **kw)
  File "C:\Users\Samuel\scoop\apps\scons\3.1.2\scons-local-3.1.2\SCons\Script\SConscript.py", line 597:
    return _SConscript(self.fs, *files, **subst_kw)
  File "C:\Users\Samuel\scoop\apps\scons\3.1.2\scons-local-3.1.2\SCons\Script\SConscript.py", line 286:
    exec(compile(scriptdata, scriptname, 'exec'), call_stack[-1].globals)
  File "C:\Users\Samuel\Dev\Godot\godot\godot\modules\SCsub", line 20:
    env.split_lib("modules", env_lib = env_modules)
  File "C:\Users\Samuel\Dev\Godot\godot\godot\methods.py", line 336:
    if '-Wl,--start-group' in env['LINKCOM'] and '-Wl,--start-group' in env['SHLINKCOM']:
```

**Steps to reproduce:**
Setup dev environment according to docs using Scoop and MinGW: https://docs.godotengine.org/en/latest/development/compiling/compiling_for_windows.html

1. Install Scoop: https://scoop.sh/
2. Install dependencies using scoop: `scoop install gcc python scons yasm make`
3. Attempt to compile project using scons: `scons -j6 platform=windows`

**Minimal reproduction project:**
Will need a fresh Windows 10 installation

**Comment by fire** [comment]
Can you post the python version on that computer? My guess is that python version changes broke the build.

**Comment by gamedevsam** [comment]
It was whatever scoop installed by default, which appears to be the latest version v3.8.1.

**Comment by akien-mga** [comment]
That's because in some conditions (don't know which, it worked for me) `env['LINKCOM']` and `env['SHLINKCOM']` may not be strings (here at least one of them appears to be a `ListAction`, which can't be used with `in`).

I guess casting them to `str()` might do the trick.

**Comment by gamedevsam** [comment]
Ok, so I changed line 336 in `methods.py` from:
`if '-Wl,--start-group' in env['LINKCOM'] and '-Wl,--start-group' in env['SHLINKCOM']:`
to
`if '-Wl,--start-group' in str(env['LINKCOM']) and '-Wl,--start-group' in str(env['SHLINKCOM']):`

I got farther in the compilation but hit another error:

```
$ scons platform=windows
scons: Reading SConscript files ...

scons: warning: No version of Visual Studio compiler found - C/C++ compilers most likely not set correctly
File "C:\Users\Samuel\Dev\Godot\godot\SConstruct", line 3, in <module>

scons: warning: No version of Visual Studio compiler found - C/C++ compilers most likely not set correctly
File "C:\Users\Samuel\Dev\Godot\godot\SConstruct", line 65, in <module>
Configuring for Windows: target=debug, bits=default
Using MinGW
Checking for C header file mntent.h... (cached) no
scons: done reading SConscript files.
scons: Building targets ...
[Initial build] Compiling ==> platform\windows\godot_windows.cpp
=====
b'g++.exe: error: /Foplatform\\windows\\godot_windows.windows.tools.64.obj: No such file or directory\ng++.exe: error: /c: No such file or directory\ng++.exe: error: /TP: No such file or directory\ng++.exe: error: /nologo: No such file or directory\ng++.exe: error: /DDEBUG_MEMORY_ALLOC: No such file or directory\ng++.exe: error: /DDISABLE_FORCED_INLINE: No such file or directory\ng++.exe: error: /DDEBUG_ENABLED: No such file or directory\ng++.exe: error: /DDEBUG_MEMORY_ENABLED: No such file or directory\ng++.exe: error: /DWINDOWS_ENABLED: No such file or directory\ng++.exe: error: /DOPENGL_ENABLED: No 
such file or directory\ng++.exe: error: /DWASAPI_ENABLED: No such file or directory\ng++.exe: error: /DWINMIDI_ENABLED: No such file or directory\ng++.exe: error: /DWINVER=0x0601: No such file or directory\ng++.exe: error: /D_WIN32_WINNT=0x0601: No such file or directory\ng++.exe: error: /DMINGW_ENABLED: No such file or directory\ng++.exe: error: /DMINGW_HAS_SECURE_API=1: No such file or directory\ng++.exe: error: /DPTRCALL_ENABLED: No such file or directory\ng++.exe: error: /DTOOLS_ENABLED: No such file or directory\ng++.exe: error: /DGDSCRIPT_ENABLED: No such file or directory\ng++.exe: error: /DMINIZIP_ENABLED: No such file or directory\ng++.exe: error: /DZSTD_STATIC_LINKING_ONLY: No such file or directory\ng++.exe: error: /DGLAD_ENABLED: No such file or directory\ng++.exe: error: /DGLES_OVER_GL: No such file or directory\ng++.exe: error: /DFREETYPE_ENABLED: No such file or directory\ng++.exe: error: /DSVG_ENABLED: No such file or directory\ng++.exe: error: /Ithirdparty\\nanosvg: No such file or directory\ng++.exe: error: /Ithirdparty\\freetype\\include: No such file or directory\ng++.exe: error: /Ithirdparty\\libpng: No such file or directory\ng++.exe: error: /Ithirdparty\\glad: No such file or 
directory\ng++.exe: error: /Ithirdparty\\zstd: No such file or directory\ng++.exe: error: /Ithirdparty\\zlib: No such file or directory\ng++.exe: error: /Iplatform\\windows: No such file or directory\ng++.exe: error: /I.: No such file or directory\n'
=====
scons: *** [platform\windows\godot_windows.windows.tools.64.obj] Error 1
scons: building terminated because of errors.
```

Any ideas?

**Comment by gamedevsam** [comment]
I want to let you all know I fixed my problem by installing Microsoft Build Tools according to these instructions (I really didn't want to install the FULL Visual Studio setup since I use VSCode exclusively): https://github.com/ponylang/ponyc/blob/master/INSTALL.md#windows

**Comment by gamedevsam** [comment]
Maybe we should update the docs with this alternative way of getting all the required dependencies without installing Visual Studio proper, and remove the part about using scoop since it's currently broken.

Ex:

```
Visual Studio 2019, 2017 or 2015 (available here) or the Visual C++ Build Tools 2019, 2017 or 2015 (available here), and
If using Visual Studio 2015, install the Windows 10 SDK (available here).
If using Visual Studio 2017 or 2019, install the "Desktop Development with C++" workload.
If using Visual C++ Build Tools 2017 or 2019, install the "Visual C++ build tools" workload, and the "Visual Studio C++ core features" individual component.
If using Visual Studio 2017 or 2019, or Visual C++ Build Tools 2017 or 2019, make sure the latest Windows 10 SDK (10.x.x.x) for Desktop will be installed.
```

**Comment by rambda** [comment]
Encountered the exactly same error when I manually installed MSYS2, MinGW and other dependencies and attempted to build Godot. I tried cast the env vars to strings too, got another error which is exactly same as OP again.
What worked for me is using scoop though.

 I used to build Godot 3.2 with the same setup, and everything was fine at that time. This year I reinstalled my Windows 10, and then this setup did not work until I tried scoop.
